tools/metrics/gauges_elasticache.go

```diff
@@ -12,60 +18,239 @@ import (
+				{Label: "replication_group_id", Value: replicationGroupId},
+				{Label: "service_instance_guid", Value: details.ServiceInstance.Guid},
+				{Label: "space_name", Value: details.Space.Name},
+				{Label: "space_guid", Value: details.Space.Guid},
+				{Label: "org_name", Value: details.Org.Name},
+				{Label: "org_guid", Value: details.Org.Guid},
+			},
 		})
+	}
 
-		return w.WriteMetrics(metrics)
+	metrics = append(metrics, m.Metric{
+		Kind:  m.Gauge,
+		Time:  time.Now(),
+		Name:  "aws.elasticache.node.count",
+		Value: float64(totalNodeCount),
+		Unit:  "count",
 	})
+	return metrics, nil
+}
+
+func fetchRedisServiceInstances(cfAPI cfclient.CloudFoundryClient) ([]CFRedisService, error) {
+	spacesCache := map[string]cfclient.Space{}
+	orgsCache := map[string]cfclient.Org{}
+
+	servicesWithRedisLabel, err := cfAPI.ListServicesByQuery(url.Values{
+		"q": []string{"label:redis"},
+	})
+
+	if err != nil {
+		return []CFRedisService{}, err
+	}
+
+	if len(servicesWithRedisLabel) == 0 {
+		return nil, fmt.Errorf("could not find service with label=redis")
+	}
+
+	redisServiceGuid := servicesWithRedisLabel[0].Guid
+	redisServicePlans, err := cfAPI.ListServicePlansByQuery(url.Values{
+		"q": []string{"service_guid:" + redisServiceGuid},
+	})
+
+	if err != nil {
+		return []CFRedisService{}, err
+	}
+
+	var serviceDetails []CFRedisService
+	for _, plan := range redisServicePlans {
+		planInstances, err := cfAPI.ListServiceInstancesByQuery(url.Values{
+			"q": []string{"service_plan_guid:" + plan.Guid},
+		})
+
+		if err != nil {
+			return []CFRedisService{}, err
+		}
+
+		for _, instance := range planInstances {
+			var space cfclient.Space
+			if s, ok := spacesCache[instance.SpaceGuid]; ok {
+				space = s
+			} else {
+				space, err = cfAPI.GetSpaceByGuid(instance.SpaceGuid)
+				if err != nil {
+					return []CFRedisService{}, err
+				}
+
+				spacesCache[space.Guid] = space
+			}
+
+			var org cfclient.Org
+			if o, ok := orgsCache[space.OrganizationGuid]; ok {
+				org = o
+			} else {
+				org, err = cfAPI.GetOrgByGuid(space.OrganizationGuid)
+				if err != nil {
+					return []CFRedisService{}, err
+				}
+
+				orgsCache[org.Guid] = org
+			}
+
+			serviceDetails = append(serviceDetails, CFRedisService{
+				ServiceInstance: instance,
+				Space:           space,
+				Org:             org,
+			})
+		}
+	}
+
+	return serviceDetails, nil
+}
+
+func iterateCacheParameterGroups(client elasticacheiface.ElastiCacheAPI, fn func(*awsec.CacheParameterGroup) error) error {
+	errs := []error{}
+	err := client.DescribeCacheParameterGroupsPages(
+		&awsec.DescribeCacheParameterGroupsInput{},
+		func(page *awsec.DescribeCacheParameterGroupsOutput, lastPage bool) bool {
+			for _, cacheParameterGroup := range page.CacheParameterGroups {
+				err := fn(cacheParameterGroup)
+				if err != nil {
+					errs = append(errs, err)
+				}
+			}
+			return true
+		},
+	)
+	if err != nil {
+		return err
+	}
+	if len(errs) > 0 {
+		return errs[0]
```

Comment:
It's not a blocker of any sort, but I'm debating if it'd be better to return all errors as oppose to single one.

I think it's good avoid having to re-run the same thing multiple times, just to step on another error?

```go
fmt.Errorf("%d errors:\n%s", len(errs), errs)

// 5 errors:
// [1: one 2: two 3 4: four 5: five]
```

would be improvement.

In order to make it useful and pretty it'd be more work. I'm happy for it to stay as is :)